packages/cli/src/shared.ts

```diff
@@ -0,0 +1,512 @@
+      task: () => {
+        const newProgram = getProgram({
+          compilerOptions: {
+            ...compilerOptions,
+            module: typescript.ModuleKind.CommonJS,
+
+            // `ModuleResolutionKind.NodeJs` is deprecated in TypeScript 5 and
+            // later, but TypeScript 4 doesn't support
+            // `ModuleResolutionKind.Node10`.
+            moduleResolution: ModuleResolutionKind.NodeJs,
+          },
+          projectReferences,
+          files,
+          oldProgram: program,
+          host,
+        });
+
+        build({
+          program: newProgram,
+          type: 'commonjs',
+          system,
+          verbose,
+          shims,
+        });
+      },
+    },
+  ];
+
+  executeSteps(buildSteps, {}, verbose);
+}
+
+/**
+ * Build the project using the Node.js 16 module resolution strategy.
+ *
+ * @param options - The build options.
+ * @param options.name - The name of the project to build.
+ * @param options.program - The TypeScript program to build.
+ * @param options.format - The formats to build.
+ * @param options.system - The file system to use.
+ * @param options.verbose - Whether to enable verbose logging.
+ * @param options.shims - Whether to generate shims for environment-specific
+ * APIs.
+ */
+export function buildNode16({
+  name,
+  program,
+  format,
+  system,
+  verbose,
+  shims,
+}: BuilderOptions) {
+  const buildSteps: Steps<Record<string, never>> = [
+    {
+      name: `Building ES module "${name}".`,
+      condition: () => format.includes('module'),
+      task: () => {
+        build({ program, type: 'module', system, shims, verbose });
+      },
+    },
+    {
+      name: `Building CommonJS module "${name}".`,
+      condition: () => format.includes('commonjs'),
+      task: () => {
+        build({ program, type: 'commonjs', system, shims, verbose });
+      },
+    },
+  ];
+
+  executeSteps(buildSteps, {}, verbose);
+}
+
+/**
+ * Build the project references. This function will build the project references
+ * using the specified formats.
+ *
+ * @param options - The build options.
+ */
+export async function buildProjectReferences(options: BuilderOptions) {
+  const {
+    name,
+    tsConfig,
+    program,
+    format,
+    baseDirectory,
+    verbose,
+    shims,
+    clean,
+  } = options;
+
+  const resolvedProjectReferences = getDefinedArray(
+    program.getResolvedProjectReferences(),
+  );
+
+  const graph = createGraph(resolvedProjectReferences);
+  const sortedProjectReferences = getResolvedProjectReferences(
+    baseDirectory,
+    resolvedProjectReferences,
+  );
+
+  const fn = getBuildWorkerFunction({
+    parentBaseDirectory: baseDirectory,
+    format,
+    verbose,
+    shims,
+    clean,
+  });
+
+  await parallelise(sortedProjectReferences, graph, fn);
+
+  verbose &&
+    info(`All project references built. Building main project "${name}".`);
+
+  const buildFunction = getBuildFunction(tsConfig, false);
```

Comment:
Just for my own education: it seems that we only call `buildProjectReferences()` once per project (if it uses project references), and then we call the single-project build functions (i.e. `node10` or `node16`) for every project thereafter, in dependency order.

_Question:_ Where do the single-project build functions fail if project references are not built in order? I assume they collect the output files of their dependents at some point, but I am struggling to identify where that is.

Reply:
It would fail when creating the TypeScript program. TypeScript verifies that all the referenced files exist, which is also why we need to provide a custom compiler host to this function which rewrites the extensions.